src/storage/src/storage/upload_object/unbuffered.rs

```diff
@@ -35,6 +36,114 @@ where
+        } else {
+            self.send_unbuffered_single_shot().await
+        }
+    }
+
+    async fn send_unbuffered_resumable(self, hint: (u64, Option<u64>)) -> Result<Object> {
+        let mut upload_url = None;
+        gax::retry_loop_internal::retry_loop(
+            async |_| self.resumable_attempt(&mut upload_url, hint).await,
+            async |duration| tokio::time::sleep(duration).await,
+            true,
+            self.options.retry_throttler.clone(),
+            Arc::new(ContinueOn308::new(self.options.retry_policy.clone())),
+            self.options.backoff_policy.clone(),
+        )
+        .await
+    }
+
+    async fn resumable_attempt(
+        &self,
+        url: &mut Option<String>,
+        hint: (u64, Option<u64>),
+    ) -> Result<Object> {
+        let offset = if let Some(upload_url) = url {
+            match self.query_resumable_upload(upload_url).await? {
+                ResumableUploadStatus::Finalized(o) => {
+                    return Ok(*o);
+                }
+                ResumableUploadStatus::Partial(o) => o,
+            }
+        } else {
+            let upload_url = self.start_resumable_upload_attempt().await?;
+            *url = Some(upload_url);
+            0_u64
+        };
+        assert!(url.is_some());
+
+        use crate::upload_source::Seek;
+        let payload = self.payload.clone();
+        payload
+            .lock()
+            .await
+            .seek(offset)
+            .await
+            .map_err(Error::ser)?;
+
+        let range = match (offset, hint.0, hint.1) {
+            (o, _, None) => format!("bytes {o}-*/*"),
+            (_, 0, Some(0)) => "bytes */0".to_string(),
+            (o, s, Some(u)) if s == u => format!("bytes {o}-{}/{s}", s - 1),
+            (o, _, Some(_)) => format!("bytes {o}-*/*"),
+        };
+        let builder = self
+            .inner
+            .client
+            .request(reqwest::Method::PUT, url.as_ref().unwrap())
+            .header("content-type", "application/octet-stream")
+            .header("Content-Range", range)
+            .header(
+                "x-goog-api-client",
+                reqwest::header::HeaderValue::from_static(&self::info::X_GOOG_API_CLIENT_HEADER),
+            );
+
+        let builder = apply_customer_supplied_encryption_headers(builder, &self.params);
+        let builder = self.inner.apply_auth_headers(builder).await?;
+        let stream = Box::pin(unfold(Some(payload), move |state| async move {
+            if let Some(payload) = state {
+                let mut guard = payload.lock().await;
+                if let Some(next) = guard.next().await {
+                    drop(guard);
+                    return Some((next, Some(payload)));
+                }
+            }
+            None
+        }));
+
+        let builder = builder.body(reqwest::Body::wrap_stream(stream));
+        let response = builder.send().await.map_err(Error::io)?;
```

Comment:
does reqwest send the stream in chunks?

Reply:
"chunk" is the most overloaded term.  I think the short answer is "no". This will send a single PUT request with all the data.

However, because the size is not known, it will use chunk-transfer encoding, but those chunks have nothing to do with the PUT transfers, which we also call chunks in places.